SDK_VERSION: add a current-sdk-version.txt file and use it for documentation and workflows

fabiobaltieri opened the pull request:
This is meant to be used for https://github.com/zephyrproject-rtos/zephyr/issues/64926 to remove the hardcoded SDK version and have an "auto" option that gets the current one in the Zephyr checkout automatically.

---

- add a `SDK_VERSION` top level file with the current SDK version
- replace all documentation occurrences with a set of automatic substitutions
  - https://builds.zephyrproject.io/zephyr/pr/67603/docs/develop/getting_started/installation_linux.html
  - https://builds.zephyrproject.io/zephyr/pr/67603/docs/develop/toolchains/zephyr_sdk.html
- set `ZEPHYR_SDK_INSTALL_DIR` automatically in the workflow specs
  - https://github.com/zephyrproject-rtos/zephyr-testing/tree/vfabio-testing-branch

fabiobaltieri committed:
SDK_VERSION: add a SDK_VERSION file

Add a SDK_VERSION file pointing to the current SDK version.
This is meant to be used by upstream and downstream CIs to automatically
determined the current version of SDK based on the upstream checked out
version, as well as the documentation references, and intended to stay
in sync with the version used in the main CI.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Is this a reasonable location? Was thinking about keeping it close to the other workflow just for convenience but it could go anywhere.

stephanosio reviewed (changes requested):
This will need to be accompanied by a complete solution replacing all hard-coded Zephyr SDK version numbers throughout the repository (i.e. in the GitHub workflows as well as in the docs).

Considering the documentation case, it would be better to completely decouple the "recommended SDK version number" from the GitHub Actions -- for example, we could introduce something like `SDK_VERSION` under the Zephyr repository root that specifies the "recommended Zephyr SDK version" for a particular Zephyr release.

gmarull reviewed (commented):
note sure if github workflows is the right folder, why not a top-level file named "SDK-VERSION"?

fabiobaltieri commented:
Top level `SDK_VERSION` sounds good to me

> This will need to be accompanied by a complete solution replacing all hard-coded Zephyr SDK version numbers throughout the repository (i.e. in the GitHub workflows as well as in the docs).
> 
> Considering the documentation case, it would be better to completely decouple the "recommended SDK version number" from the GitHub Actions

Cool, yeah the actions could read the file and put it back in the environment and the docs could generate a substitution from it. The two statements seems to be in contrast though, are you suggesting that a single file should update actions and documentation (which sounds reasonable to me) or that they should be decoupled and use two different files or something?

stephanosio commented:
> are you suggesting that a single file should update actions and documentation

Single top-level `SDK_VERSION` for both GHA and docs.

nashif commented:
> Single top-level `SDK_VERSION` for both GHA and docs.

+1

fabiobaltieri committed:
doc: use substitutions to replace all current harcoded SDK versions

This creates a set of substitution rules to replace all current usages
of SDK versions in the documentation, apart from a few that were not
meant to be copy-pastable anyway and the version has just been swapped
with a <version> placeholder.

Since code-block does not parse the content, these have all been
replaced with parsed-literal. That one though parses URLs, which cannot
be broken, so a series of URL substitutions are provided too.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
workflows: find the SDK version from the top level SDK_VERSION file

Drop all hardcoded ZEPHYR_SDK_INSTALL_DIR settings, set the variable
automatically from the SDK_VERSION file instead.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Alright all references to 0.16.4 apart from the new file are gone.

Had to change the documentation blocks from `code-block` to `parsed-literal` and add way more substitutions that I originally thought to make URL parse correctly but it works now.

The workflows have been updated as well, most are using the current config but the codesize one is not and it [seems to work fine](https://github.com/zephyrproject-rtos/zephyr/actions/runs/7532844141/job/20504235633?pr=67603), plus I pushed it on the [testing repo](https://github.com/zephyrproject-rtos/zephyr-testing/tree/vfabio-testing-branch).

stephanosio reviewed (commented):
The changes themselves look good to me; but, the first commit title seems a little off.

> [SDK_VERSION: add a **current-sdk-version.txt** file](https://github.com/zephyrproject-rtos/zephyr/pull/67603/commits/55527fd98666f5d57bd8a53880c587eebb85e981)

carlescufi reviewed (dismissed):
Very nice, thanks Fabio!

kartben commented on doc/develop/getting_started/installation_linux.rst:
```diff
@@ -227,32 +227,31 @@ The Zephyr SDK supports the following target architectures:
+#. Download and verify the `Zephyr SDK bundle`_:
 
-   .. code-block:: bash
+   .. parsed-literal::
```
Nice find! I actually wanted to implement substitutions for the SDK version a while back but couldn't find a way to have them work in code-blocks :)
One thing that I'd like to see is that the block still renders the same as other code blocks (ex. it should have a background ; ideally, syntax highlighting too, I guess) 

<img width="800" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/ed6a52ca-b60f-45a0-a87b-c7e93512a478">

kartben replied:
https://github.com/sphinx-doc/sphinx/issues/3306

fabiobaltieri replied:
Yeah, it's a bit of a compromise, I've figured you'd probably follow up with a strategic CSS change.

kartben replied:
I don't want to block the rest of this PR, but would it make sense to keep the doc update for later? Ideally you'd open a separate PR with the doc commit there that we (I) can tackle in due course. Beyond the (arguably not critical) issue described in this comment, there is a more concerning one with... wait for it... the PDF file :)

<img width="729" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/2bab5d53-0fd2-4e3c-9098-03c0c4430e40">

fabiobaltieri replied:
> the PDF file

I HAVE AN IDEA!

fabiobaltieri replied:
...is there a way of making the `parsed-literal` not highlight? Vaguely remember reading something about it, I'll look a bit and if that's not feasible I can pull the doc change.

gmarull replied:
> > the PDF file
> 
> I HAVE AN IDEA!

PLEASE KILL PDF ONCE AND FOR ALL

fabiobaltieri commented:
@kartben I've added some backticks to the replaced string, can you give it a try? It removes highlighting on the html version, who knows what'll do in the pdf...

fabiobaltieri commented:
ha, seems like it worked?
![x](https://github.com/zephyrproject-rtos/zephyr/assets/891546/61e3db05-701d-48ea-8e10-482ca99482ab)
